Reforzar sistema completo de autenticación JWT

21Enzo17 opened the pull request:
Implementación de múltiples mejoras de seguridad en el sistema de autenticación:

1. Refresh Tokens:
   - Crear entidad RefreshToken usando UUID como identificador para mayor seguridad
   - Implementar sistema completo de gestión de refresh tokens (servicio, repo, pruebas)
   - Modificar AuthService para soportar renovación de tokens sin re-autenticación

2. Seguridad JWT mejorada:
   - Actualizar algoritmo de firma de HS256 a RS256 usando par de claves asimétricas
   - Implementar rotación periódica automática de claves de firma
   - Reducir tiempo de expiración de tokens de acceso (15 minutos)
   - Añadir validación de claims adicionales (issuer, audience, nonce)

3. Lista negra de tokens:
   - Implementar sistema Redis para almacenamiento eficiente de tokens revocados
   - Configurar expiración automática de tokens en Redis según su fecha de caducidad
   - Añadir comprobación en JwtAuthorizationFilter para validar contra lista negra

4. Mantenimiento automático:
   - Programar tareas para limpieza periódica de tokens expirados y revocados
   - Añadir monitoreo y registro detallado de operaciones críticas de seguridad

Estas mejoras protegen contra múltiples vulnerabilidades:
- Robo de tokens (tiempo corto de vida + lista negra)
- Ataques por fuerza bruta (algoritmos seguros + claves rotativas)
- Reutilización de tokens (verificación mejorada + lista negra)
- Ataques de reproducción (validación completa de claims)

Se incluyen pruebas unitarias e integración para validar todas las nuevas funcionalidades.

21Enzo17 committed:
Reforzar sistema completo de autenticación JWT

Implementación de múltiples mejoras de seguridad en el sistema de autenticación:

1. Refresh Tokens:
   - Crear entidad RefreshToken usando UUID como identificador para mayor seguridad
   - Implementar sistema completo de gestión de refresh tokens (servicio, repo, pruebas)
   - Modificar AuthService para soportar renovación de tokens sin re-autenticación

2. Seguridad JWT mejorada:
   - Actualizar algoritmo de firma de HS256 a RS256 usando par de claves asimétricas
   - Implementar rotación periódica automática de claves de firma
   - Reducir tiempo de expiración de tokens de acceso (15 minutos)
   - Añadir validación de claims adicionales (issuer, audience, nonce)

3. Lista negra de tokens:
   - Implementar sistema Redis para almacenamiento eficiente de tokens revocados
   - Configurar expiración automática de tokens en Redis según su fecha de caducidad
   - Añadir comprobación en JwtAuthorizationFilter para validar contra lista negra

4. Mantenimiento automático:
   - Programar tareas para limpieza periódica de tokens expirados y revocados
   - Añadir monitoreo y registro detallado de operaciones críticas de seguridad

Estas mejoras protegen contra múltiples vulnerabilidades:
- Robo de tokens (tiempo corto de vida + lista negra)
- Ataques por fuerza bruta (algoritmos seguros + claves rotativas)
- Reutilización de tokens (verificación mejorada + lista negra)
- Ataques de reproducción (validación completa de claims)

Se incluyen pruebas unitarias e integración para validar todas las nuevas funcionalidades.

21Enzo17 committed:
Implementación completa de mejoras de seguridad en JWT

Implementa todas las mejoras de seguridad planificadas para el sistema de autenticación:

- Migración de algoritmo HS512 a RS256 con criptografía asimétrica:
  * Generación y almacenamiento de claves RSA de 2048 bits
  * Firma de tokens usando clave privada y verificación con clave pública
  * Manejo seguro de rutas de almacenamiento de claves

- Implementación de rotación periódica automática de claves:
  * Tarea programada mensual para renovación de par de claves
  * Registro de resultados de rotación en logs del sistema

- Mejoras en la seguridad de tokens:
  * Reducción del tiempo de expiración de 30 a 15 minutos
  * Adición de nonce aleatorio para prevenir ataques de repetición
  * Validación mejorada de claims (issuer, audience)

- Pruebas unitarias completas:
  * Verificación del algoritmo RS256 en tokens generados
  * Validación de rotación de claves y su persistencia
  * Comprobación de nonce y tiempo de expiración correcto
  * Pruebas para tareas programadas de mantenimiento

Esta implementación mejora significativamente la seguridad del sistema contra
múltiples vectores de ataque, incluyendo robos de tokens, ataques por
fuerza bruta y ataques de reproducción.

21Enzo17 committed:
Añadir documentación detallada del sistema de seguridad

Añade documentación técnica completa del sistema de seguridad:

- Nuevo archivo docs/SEGURIDAD.md con documentación detallada sobre:
  * Arquitectura de seguridad basada en JWT y RS256
  * Sistema de refresh tokens y gestión de sesiones
  * Lista negra de tokens con Redis
  * Rotación automática de claves criptográficas
  * Protección contra vectores de ataque comunes

- Actualización del README.md principal:
  * Nueva sección sobre las características de seguridad
  * Referencia a la documentación detallada
  * Badge de Redis añadido a tecnologías utilizadas

Esta documentación proporciona una referencia técnica completa del sistema
de seguridad para facilitar la mantenibilidad, desarrollo futuro y auditorías
de seguridad.

21Enzo17 committed:
Se agrega trim a la cadena de JWT, por sugerencia copilot

DarioRv commented on config/keys/jwt_public.key:
No subir secret key, aunque sean de prueba

DarioRv commented on config/keys/jwt_private.key:
No subir secret key, aunque sean de prueba

DarioRv reviewed (changes requested):
Eliminar archivos que exponen secret keys

DarioRv commented on .env:
Estas variables deben verse reflejadas en `application.properties`, o bien migrar la variables de `application.properties` para utilizar `.env`

DarioRv commented on pom.xml:
> Estas variables deben verse reflejadas en `application.properties`, o bien migrar la variables de `application.properties` > para utilizar `.env`

Si no se opta por migrar a .env entonces eliminar dependencia `dotenv-java`

DarioRv commented on docs/SEGURIDAD.md:
````diff
@@ -0,0 +1,319 @@
+}
+```
+
+## Rotación de Claves Criptográficas
````
Esto implica que los tokens firmados con una antigua clave quedé inválido, significando el cierre de sesión para todos los usuarios el primer día del mes. ¿Esto es así? ¿Es combieniente este planteo?

DarioRv reviewed (commented):
Revisar los comentarios.

21Enzo17 committed:
feat(security): implementar sistema robusto de rotación de claves JWT y mejorar gestión de tokens

Implementación de un sistema avanzado de rotación y archivado de claves JWT para mejorar la seguridad de la autenticación:

- Refactor completo de JwtUtil para soportar rotación automática de claves RSA (2048 bits)
- Implementación de sistema de archivado de claves públicas anteriores con límite configurable
- Validación continua de tokens firmados con claves anteriores durante período de transición
- Carga dinámica de claves públicas archivadas para validación de tokens existentes
- Eliminación automática de claves antiguas que excedan el límite configurado
- Mejora de la validación de tokens con verificación completa de issuer, audience y claims
- Implementación de nonce único por token para prevenir ataques de repetición
- Adición del método existsByToken a RefreshTokenService para verificación eficiente
- Migración de configuración hardcodeada a variables de entorno configurables mediante .env
- Creación de clase EnvConfig para gestionar la carga de variables desde archivo .env
- Actualización de pruebas unitarias con H2 y Redis embebido para simulación completa
- Creación de TestSecurityConfig para pruebas integradas de autenticación
- Implementación de JwtKeyRotationTest para validar rotación y preservación de tokens
- Creación de archivo .env.example como referencia para configuración en producción
- Actualización del .gitignore para excluir archivos sensibles (claves, credenciales)
- Actualización de documentación en SEGURIDAD.md con detalles del nuevo sistema
- Configuración de variables de entorno para pruebas en application-test.properties

Esta implementación refuerza significativamente la seguridad del sistema mediante la rotación periódica de material criptográfico, minimizando los riesgos asociados con la exposición de claves a largo plazo, mientras mantiene la compatibilidad con tokens existentes durante los períodos de transición.

21Enzo17 committed:
Iimplementar sistema robusto de rotación de claves JWT y mejorar gestión de tokens

Implementación de un sistema avanzado de rotación y archivado de claves JWT para mejorar la seguridad de la autenticación:

- Refactor completo de JwtUtil para soportar rotación automática de claves RSA (2048 bits)
- Implementación de sistema de archivado de claves públicas anteriores con límite configurable
- Validación continua de tokens firmados con claves anteriores durante período de transición
- Carga dinámica de claves públicas archivadas para validación de tokens existentes
- Eliminación automática de claves antiguas que excedan el límite configurado
- Mejora de la validación de tokens con verificación completa de issuer, audience y claims
- Implementación de nonce único por token para prevenir ataques de repetición
- Adición del método existsByToken a RefreshTokenService para verificación eficiente
- Migración de configuración hardcodeada a variables de entorno configurables mediante .env
- Creación de clase EnvConfig para gestionar la carga de variables desde archivo .env
- Actualización de pruebas unitarias con H2 y Redis embebido para simulación completa
- Creación de TestSecurityConfig para pruebas integradas de autenticación
- Implementación de JwtKeyRotationTest para validar rotación y preservación de tokens
- Creación de archivo .env.example como referencia para configuración en producción
- Actualización del .gitignore para excluir archivos sensibles (claves, credenciales)
- Actualización de documentación en SEGURIDAD.md con detalles del nuevo sistema
- Configuración de variables de entorno para pruebas en application-test.properties

Esta implementación refuerza significativamente la seguridad del sistema mediante la rotación periódica de material criptográfico, minimizando los riesgos asociados con la exposición de claves a largo plazo, mientras mantiene la compatibilidad con tokens existentes durante los períodos de transición.

21Enzo17 committed:
fix(auth): Estandarizar códigos HTTP para errores de autenticación

Unifica el uso de códigos de estado HTTP en JwtAuthorizationFilter para
mantener consistencia en la respuesta de errores de autenticación:

- Cambia el código HTTP de FORBIDDEN (403) a UNAUTHORIZED (401) para todos
  los errores relacionados con tokens JWT
- Mejora la coherencia de la API al usar el mismo código para el mismo
  tipo de error (problemas de autenticación)
- Se alinea con las mejores prácticas REST donde 401 se usa para errores
  de credenciales y 403 para problemas de permisos

Este cambio aumenta la predictibilidad de las respuestas de error para
los consumidores de la API.

21Enzo17 committed:
refactor(auth): simplificar método sendErrorResponse en JwtAuthorizationFilter

Elimina el parámetro 'details' no utilizado del método sendErrorResponse para
mejorar la claridad y mantenibilidad del código:

- Elimina el parámetro 'details' que no se usaba en la creación del ApiResponse
- Actualiza las llamadas al método en los bloques catch para usar la nueva firma
- Mantiene el registro detallado de errores en los logs para depuración
- Sigue las buenas prácticas de seguridad al no exponer detalles técnicos
  de errores de autenticación al frontend

Este cambio hace que el código sea más limpio y coherente con el enfoque
actual de seguridad de la aplicación.

DarioRv reviewed (commented):
Añadir dockerfile

21Enzo17 committed:
feat(auth): implementar funcionalidad de renovación de sesión completa con refresh token

- Añadido nuevo endpoint /api/v1/auth/renovar-sesion para restaurar sesiones de usuario
- Creado método renovarSesion en IAuthService y su implementación
- Optimizado el flujo de autenticación para mejorar la experiencia de usuario tras cierre de navegador
- Se obtienen datos del usuario, nuevo token y refresh token en una sola petición

Este cambio permite recuperar la sesión completa del usuario usando únicamente el refresh token guardado, mejorando la seguridad al no requerir almacenar datos sensibles en el cliente.

21Enzo17 committed:
Cambios el en readme de seguridad
